### PR TITLE
DSLTree

### PR DESCRIPTION
**Pull request by milseman**
DSLTree is an internal representation shared between literals and result builders. 

- Lowering/transforming AST -> DSLTree
  - TODO: More atoms, built-in character classes, etc
- Pretty-printing DSLTree
  - At rough parity of the original PrintAsPattern
- Switch compiler over to DSLTree
- Switch result builders over to DSLTree
- Legacy VM support for everything except the new extensions
- CaptureStructure support

### CODE REVIEW

**Review by rxwei — APPROVED** [review]

File: Sources/_MatchingEngine/Regex/DSLTree.swift [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+            result += str
+            astChildren.formIndex(after: &idx)
+          }
+          return result.isEmpty ? nil : result
+        }
+
+        // No need to nest single children concatenations
+        if astChildren.count == 1 {
+          return astChildren.first!.dslTreeNode
+        }
```
Comment (hamishknight) [inline Sources/_MatchingEngine/Regex/DSLTree.swift:200]: When do single-child concatenations occur? I kind of wonder whether `AST.Concatenation` should enforce that `children.count >= 2` (and maybe have a utility that produces concat-or-self), or is it better for compatibility with the DSL to allow them?

**Review by hamishknight — APPROVED** [review]

File: Sources/_MatchingEngine/Regex/DSLTree.swift (hunk 1 above)
Comment (rxwei) [reply Sources/_MatchingEngine/Regex/DSLTree.swift:200]: I’m in favor of enforcing a plural count.

Comment (milseman) [reply Sources/_MatchingEngine/Regex/DSLTree.swift:200]: How would we enforce that in the DSL? Would we error out if given a single member? I worry there could be impact with sensible refactoring like quoting a literal run of characters vs listing each separately.

Comment (rxwei) [reply Sources/_MatchingEngine/Regex/DSLTree.swift]: The builder DSL always supports having a singular element in a block, e.g. `Regex { "a" }`. It's just that `buildBlock(_:)` (unary) never creates a concatenation in this scenario, but returns the argument instead.  As such, I'm not sure when else a singular concatenation would be created.

File: Sources/_MatchingEngine/Regex/DSLTree.swift [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    case trivia(String)
+
+    // TODO: Probably some atoms, built-ins, etc.
+
+    case empty
+
+    case stringLiteral(String)
+
+    /// An embedded literal
+    case regexLiteral(AST.Node)
```
[the remark this answers is not in the document]
Comment (hamishknight) [reply Sources/_MatchingEngine/Regex/DSLTree.swift]: Hmm yeah, I guess they would have to form their own scope in which they're only active in that portion of the regex

**Comment by milseman** [comment]
@swift-ci please test linux platform

**Comment by milseman** [comment]
@hamishknight @rxwei I had to move some stuff around, can you take a look at this PR again to see if it's mergeable?

**Review by rxwei — APPROVED** [review]

**Pull request merged**